File: src/object.c
//
//  object.c
//  CObjects
//
//  Created by George Boumis on 9/2/13.
//  Copyright (c) 2013 George Boumis. All rights reserved.
//

#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <pthread.h>
#include <errno.h>
#include <memory_management/memory_management.h>


#include <coassert.h>
#include <compilerDefs.h>
#include <Object.h>
#include <Object.r>
#include <new.h>
#include <StringObject.h>
#include <AutoreleasePool.h>
#undef retain
#undef release


void * constructor(void * self, va_list * app) {
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,ENOTSUP,NULL);
	return class->constructor(self, app);
}

void * destructor(void * self) {
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,NULL);
	COAssertNoNullOrReturn(class->destructor,ENOTSUP,NULL);
	return class->destructor(self);
}

void * super_constructor(const void *const class, void * self, va_list * app) {
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	COAssertNoNullOrReturn(class,EINVAL,NULL);
	
	const struct Classs * _superclass = superclass(class);
	COAssertNoNullOrReturn(_superclass,EINVAL,NULL);
	COAssertNoNullOrReturn(_superclass->constructor,ENOTSUP,NULL);
	return _superclass->constructor(self, app);
}

void * super_destructor(const void *const class, void * self) {
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	COAssertNoNullOrReturn(class,EINVAL,NULL);
	
	const struct Classs * _superclass = superclass(class);
	COAssertNoNullOrReturn(_superclass,EINVAL,NULL);
	COAssertNoNullOrReturn(_superclass->destructor,ENOTSUP,NULL);
	return _superclass->destructor(self);
}


//***********//


void * Object_constructor (void * _self, va_list * app) {
	struct Object *self = _self;
	MEMORY_MANAGEMENT_ATTRIBUTE_SET_DEALLOC_FUNCTION(self, delete);
	return self;
}

void * Object_destructor (void * _self) {
	return _self;
}

void * Object_copy (const void * const _self) {
	return new(Object);
}

bool Object_equals (const void * const _self, const void *const _other) {
	const struct Object *const self = _self;
	const struct Object *const other = _other;
	return (bool)(self == other);
}

//GCC_DIAG_ON(no-pointer-to-int-cast)
UInteger Object_hash (const void *const _self) {
	const struct Object *const self = _self;
	UInteger prime = 31;
	UInteger result = 1;
	result = prime * result + (UInteger)self;
	return result;
}
//GCC_DIAG_OFF(no-pointer-to-int-cast)

void * Object_retain (void * const _self) {
	struct Object *const self = _self;
	MEMORY_MANAGEMENT_RETAIN(self);
	return self;
}

void Object_release (void * const _self) {
	struct Object *const self = _self;
	MEMORY_MANAGEMENT_RELEASE(self);
}

UInteger Object_retainCount (const void * const _self) {
	const struct Object *const self = _self;
	return MEMORY_MANAGEMENT_GET_RETAIN_COUNT(self);
}

void * Object_autorelease (void * _self) {
	return _self;
}

StringRef Object_copyDescription (const void * const _self) {
	const struct Object *const self = _self;
	const struct Classs *class = classOf(self);
	StringRef copyDescription = newStringWithFormat(String, "<%s %p retainCount:%ul>", class->class_name, self, retainCount(self), NULL);
	return copyDescription;
}


/* */

const char *getClassName(const void *const self) {
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,NULL);
	return class->class_name;
}

void * copy(const void *const self) {
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,NULL);
	COAssertNoNullOrReturn(class->copy,ENOTSUP,NULL);
	
	return class->copy(self);
}

bool equals(const void *const self, const void *const other) {
	COAssertNoNullOrReturn(self,EINVAL,-1);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,-1);
	COAssertNoNullOrReturn(class->equals,ENOTSUP,-1);
	
	return class->equals(self, other);
}

UInteger hash (const void *const self) {
	COAssertNoNullOrReturn(self,EINVAL,0);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,0);
	COAssertNoNullOrReturn(class->hash,ENOTSUP,0);
	
	return class->hash(self);
}

void * retain (void *const self) {
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,NULL);
	COAssertNoNullOrReturn(class->retain,ENOTSUP,NULL);
	
	return class->retain(self);
}

void release (void *const self) {
	COAssertNoNullOrBailOut(self,EINVAL);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrBailOut(class,EINVAL);
	COAssertNoNullOrBailOut(class->release,ENOTSUP);
	
	class->release(self);
}

UInteger retainCount (const void *const self) {
	COAssertNoNullOrReturn(self,EINVAL,0);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,0);
	COAssertNoNullOrReturn(class->retainCount,ENOTSUP,0);
	
	return class->retainCount(self);
}

bool instanceOf (const void * const self, const void *const _class) {
	COAssertNoNullOrReturn(self,EINVAL,-1);
	COAssertNoNullOrReturn(_class,EINVAL,-1);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,-1);
	
	/* Get the type */
	const struct Classs *const mclass = _class;
	return ( strcmp(class->class_name, mclass->class_name) == 0 );
}

bool isSubclassOf (const void *const self, const void * const _class) {
	bool result = instanceOf(self, _class);
	if ( !result )
		result = isSubclassOf(superclass(self), _class);
	return result;
}

UInteger sizeOf(const void *const self) {
	COAssertNoNullOrReturn(self,EINVAL,0);
	
	const struct Classs *const class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,0);
	
	return class->size;
}

const void * classOf(const void * const _self) {
	COAssertNoNullOrReturn(_self,EINVAL,NULL);
	
	const struct Object *const self = _self;
	COAssertNoNullOrReturn(self->class,EINVAL,NULL);
	return (const void *)self->class;
}

const void * superclass (const void *const _class) {
	COAssertNoNullOrReturn(_class,EINVAL,NULL);
	
	const struct Classs * self = _class;
	COAssertNoNullOrReturn(self->super,EINVAL,NULL);
	
	return self->super;
}

const void * super(const void *const _self) {
	COAssertNoNullOrReturn(_self,EINVAL,NULL);
	
	const struct Classs * self = classOf(_self);
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	COAssertNoNullOrReturn(self->super,ENOTSUP,NULL);
	
	return (const void *)self->super;
}

void * copyDescription (const void *const self) {
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	
	const struct Classs *class = classOf(self);
	COAssertNoNullOrReturn(class,EINVAL,NULL);
	COAssertNoNullOrReturn(class->copyDescription,ENOTSUP,NULL);
	
	return class->copyDescription(self);
}


void *autorelease (void *const self) {
	COAssertNoNullOrReturn(self,EINVAL,NULL);
	AutoreleasePoolAddObject(self);
	return self;
//	assert(self!=NULL);
//	if ( self == NULL ) return errno = EINVAL, NULL;
//	
//	const struct Classs *class = classOf(self);
//	assert(class->autorelease != NULL);
//	if ( class->autorelease == NULL ) return errno = ENOTSUP, NULL;
//	
//	return class->autorelease(self);
}

File: include/Object.r
/*!
//  @file Object.r
//  CObjects
//
//  Created by George Boumis on 9/2/13.
//  @Copyright (c) 2013 George Boumis. All rights reserved.
*/

#ifndef CObjects_object_r
#define CObjects_object_r

#include <pthread.h>

/*!
 *  @class Object Object.r
 *  @private
 *  @brief The basic structure/class that describes an @ref Object type.
 *  @details This class contains a pointer to a @ref Class instance, a reference counter, and a @a pthread_mutex_t for internal purposes.
 */
struct Object {
	/*!
	 *  @member class
	 *  @protected
	 *  @brief A pointer to a @ref Class describing this instances behaviour.
	 */
	const struct Classs * class;
	/*!
	 *  @member reatainCount
	 *  @protected
	 *  @brief The reference counter of this instance.
	 */
	//unsigned long retainCount;
	/*!
	 *  @member guardian
	 *  @protected
	 *  @brief Used for thread safe reference counter operations.
	 */
	//pthread_mutex_t guardian;
};

/*!
 *  @typedef typedef void * (* ctor) (void * self, va_list * app)
 *  @private
 *  @brief The form of an @ref Object @ref constructor.
 */
typedef void * (* ctor) (void * self, va_list * app);

/*!
 *  @typedef typedef void * (* dtor) (void * self)
 *  @private
 *  @brief The form of an @ref Object @ref destructor.
 */
typedef void * (* dtor) (void * self);

/*!
 *  @typedef typedef typedef void * (* cpy) (const void *const self)
 *  @private
 *  @brief The form of an @ref Object's @ref copy method.
 */
typedef void * (* cpy) (const void *const self);

/*!
 *  @typedef typedef int (* eq) (const void * self, const void * other)
 *  @private
 *  @brief The form of an @ref Object's @ref equals method.
 */
typedef int (* eq) (const void * self, const void * other);

/*!
 *  @typedef typedef int (* hsh) (const void *const self)
 *  @private
 *  @brief The form of an @ref Object's @ref hash method.
 */
typedef int (* hsh) (const void *const self);

/*!
 *  @typedef typedef void * (* desc) (const void *const self)
 *  @private
 *  @brief The form of an @ref Object's @ref copyDescription method
 */
typedef void * (* desc) (const void *const self);

/*!
 *  @typedef typedef void * (* ret) (void *const self)
 *  @private
 *  @brief The form of an @ref Object's @ref retain method
 */
typedef void * (* ret) (void *const self);

/*!
 *  @typedef typedef void (* rel) (void *const self)
 *  @private
 *  @brief The form of an @ref Object's @ref release method
 */
typedef void (* rel) (void *const self);

/*!
 *  @typedef typedef unsigned long (* retc) (const void *const self)
 *  @private
 *  @brief The form of an @ref Object's @ref retainCount method
 */
typedef unsigned long (* retc) (const void *const self);


typedef void *(*autor) (void *const self);

/*!
 *  @class Class Object.r
 *  @private
 *  @brief The basic structure/class that describes an @ref Class type.
 *  @details The instance of this class is an @ref Object. This class contains the @ref Object structure, the name of the class as a @a C string, a pointer to the super class, the size of an instance of this class. It contains also functions pointers to the class's methods.
 */
struct Classs {
	/*!
	 *  @member isa
	 *  @protected
	 *  @brief The object type this class.
	 */
	struct Object isa;
	
	/*!
	 *  @member class_name
	 *  @protected
	 *  @brief The class's name as a @a C string.
	 */
	const char *class_name;
	
	/*!
	 *  @member super
	 *  @protected
	 *  @brief The pointer to the  super class's @ref Class.
	 */
	const struct Classs * super;
	
	/*!
	 *  @member size
	 *  @protected
	 *  @brief The size for this class' instances.
	 */
	size_t size;
	
	/* 
	 Methods 
	 */
	
	/*!
	 *  @member constructor
	 *  @protected
	 *  @brief The function pointer to the @ref constructor of this class.
	 */
	ctor constructor;
	
	/*!
	 *  @member destructor
	 *  @protected
	 *  @brief The function pointer to the @ref destructor of this class.
	 */
	dtor destructor;

	/*!
	 *  @member copy
	 *  @protected
	 *  @brief The function pointer to the @ref copy method of this class.
	 */
	cpy copy;
	
	/*!
	 *  @member copy
	 *  @protected
	 *  @brief The function pointer to the @ref equals method of this class.
	 */
	eq equals;

	/*!
	 *  @member hash
	 *  @protected
	 *  @brief The function pointer to the @ref hash method of this class.
	 */
	hsh hash;
	
	/*!
	 *  @member copyDescription
	 *  @protected
	 *  @brief The function pointer to the @ref copyDescription method of this class.
	 */
	desc copyDescription;
	
	
	/*
	 Memory management 
	 */
	
	/*!
	 *  @member retain
	 *  @protected
	 *  @brief The function pointer to the @ref retain method of this class.
	 */
	ret retain;
	
	/*!
	 *  @member release
	 *  @protected
	 *  @brief The function pointer to the @ref release method of this class.
	 */
	rel release;
	
	/*!
	 *  @member retainCount
	 *  @protected
	 *  @brief The function pointer to the @ref retainCount method of this class.
	 */
	retc retainCount;
	
	/*!
	 *  @member autorelease
	 *  @protected
	 *  @brief The function pointer to the @ref autorelease method of this class.
	 */
	autor autorelease;
};

/*!
 *  @protected
 *  @method void * super_constructor(const void * class, void * self, va_list * app)
 *  @relates Object
 *  @brief A method that calls the super class's constructor of a given class.
 *  @param[in] class the @ref Class which its super constructor will be called.
 *  @param[in] self the instance, of type @ref Object.
 *  @param[in] app the variadic list of arguments.
 *  @return a @ref ObjectRef pointer (probably) pointing to self.
 */
void * super_constructor(const void * class, void * self, va_list * app);

/*!
 *  @protected
 *  @method void * super_destructor(const void * class)
 *  @relates Object
 *  @brief A method that calls the super class's destructor.
 *  @param[in] self the instance of type @ref Object.
 *  @return a @ref ObjectRef pointer (probably) pointing to self.
 */
void * super_destructor(const void * class, void * self);

/*!
 *  @protected
 *  @method void * constructor(void * self, va_list * app)
 *  @relates Object
 *  @brief An @ref Object @ref constructor. It's called by @ref new() to initialize a  newly created instance. This method can return another pointer than @a self. The return value can be @a NULL if initialization failed.
 *  @param[in] self the instance of type @ref Object.
 *  @param[in] app the variadic list of arguments.
 *  @return a @ref ObjectRef pointer (probably) pointing to self.
 */
void * constructor(void * self, va_list * app);

/*!
 *  @protected
 *  @method void * super_destructor(const void * class)
 *  @relates Object
 *  @brief An @ref Object @ref destructor. It's called by @ref delete() when destroying an instance. This method should free all memory allocated by the instance. Either @a NULL or a pointer suitable to be passed to free() is returned. An instance can return @a NULL and not deallocate any memory if it wants to stay alive.
 *  @param[in] self the instance of type @ref Object.
 *  @return a @ref ObjectRef pointer (probably) pointing to self.
 */
void * destructor(void * self);



void * Object_constructor (void * _self, va_list * app);
void * Object_destructor (void * _self);
void * Object_copy (const void * const _self);
int Object_equals (const void *const _self, const void *const _other);

int Object_hash (const void *const _self);

/* Memory management */
void * Object_retain (void *const _self);
void Object_release (void *const _self);
unsigned long Object_retainCount (const void *const _self);
void * Object_autorelease (void * _self);

void * Object_copyDescription (const void *const _self);

const void * super (const void *const _self);

#endif
